### PR TITLE
Feature/offband realm #3146

### PR DESCRIPTION
**Pull request by janherich**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

Tests #3146 

### Summary:

This PR implements off-band realm reads + writes for message receive path, achieved by reading everything necessary at the startup time (when chats are loaded) + putting every relevant realm write (messages, pending-messages, processed-messages, chats) into `core.async` queue.

### Testing notes (optional):
Basically regression testing of the whole chat functionality (receiving/sending) is necessary.
Other things like wallet/profile etc. should not be affected

status: ready

### CODE REVIEW

File: src/status_im/chat/events.cljs [hunk 1]
```diff
@@ -1,9 +1,11 @@
 (ns status-im.chat.events
-  (:require [cljs.core.async :as async]
+  (:require [clojure.set :as s]
```
Comment (flexsurfer) [inline src/status_im/chat/events.cljs]: set not used ?, remove or use `set` alias

File: src/status_im/chat/events.cljs [hunk 2]
```diff
@@ -165,14 +181,20 @@
 (handlers/register-handler-fx
   :initialize-chats
   [(re-frame/inject-cofx :all-stored-chats)
```
Comment (flexsurfer) [inline src/status_im/chat/events.cljs:198]: it would be good to have macro or function something like `(inject-cofxs [:cofx1 :cofx2 :cofx3])` but not related to this PR

**Review by flexsurfer — APPROVED** [review]
Great work!

**Review by dmitryn — APPROVED** [review]

**Comment by dmitryn** [comment]
@janherich i just tested this while receiving a lot of messages from #status public channel - i got UI unresponsive again (same story with bunch of `send-seen!` events getting processed sequentially)

**Comment by janherich** [comment]
Hi @dmitryn, just to be sure, did you test the develop build, or production one from diawi with `info` log level ?
Branch: status-im:feature/offband-realm-#3146
Android: https://i.diawi.com/Y6uoPz
iOS: https://i.diawi.com/P6GL5w
The thing is, with spec checking + log lines everywhere, it's not possible to really push execution times down enough.
But with prod diawi build, I was not able to simulate blocking whatsoever, at least on iPhone SE (will try on android 6.x later today).

**Comment by dmitryn** [comment]
@janherich hi! i was using develop build, let me try https://i.diawi.com/Y6uoPz on real android

**Comment by dmitryn** [comment]
Still having lags after joining #status channel and re-opening the app. Also i got this:

screenshot_20180129-203842

File: src/status_im/chat/events.cljs (hunk 2 above)
Comment (rasom) [reply src/status_im/chat/events.cljs:198]: like `(map re-frame/inject-cofx [:cofx1 :cofx2 :cofx3])`? :)

**Comment by janherich** [comment]
@dmitryn The contacts error is already fixed, investigating perf issues on android

File: src/status_im/chat/handlers/send_message.cljs [hunk 3]
```diff
@@ -93,13 +93,13 @@
...
                             chat-id         :chat-id
                             :as             params}]]
-      (let [clock-value   (messages/get-last-clock-value chat-id)
-            request       (:request handler-data)
-            hidden-params (->> (:params command)
-                               (filter :hidden)
-                               (map :name))
-            command'      (-> (prepare-command current-public-key chat-id clock-value request content)
-                              (assoc :message-type (message-type (get chats chat-id))))]
+      (let [request       (:request handler-data)
```
Comment (rasom) [inline src/status_im/chat/handlers/send_message.cljs]: alignment

**Review by rasom — APPROVED** [review]

**Comment by janherich** [comment]
Some investigation on the problem when really lot of messages are being received -> I was now able to simulate that even on iPhone, when I joined `#status` channel earlier today, I did received ~50 messages, now hundreds of messages are received literally in matter of seconds.
But the problem is not `re-frame` queue being overwhelmed -> notice that as long as you don’t enter the chat being flooded with messages, everything (navigating to other chats, working with wallet, etc.) works totally fine without slightest lags, which it wouldn’t with re-frame queue being occupied.
I also tried to time the big `model/receive-message` function, computing all the necessary effects for new message, and it takes 1ms in dev mode (likely faster with prod build).

The problem of chat being laggy and unresponsive when flooded with a lot of messages is caused by the “new message” appearing animation - each new message is animated so it takes ~0.3s for it to slide from the bottom of the screen and “pop” in place. And this animation is backed by internal state (`r/atom`) + lot of communication through RN bridge (@roman already profiled that). Now imagine what happens when hundreds of messages appear in very short time interval and all of them go through such animation :slightly_smiling_face:

I think that the solution is to introduce some receive buffering (actually we already have that in form of `pre-receiver` to correctly sort out messages according to their clock values) and only animate the very last message in batch, so let’s say with 1s buffering, we will make sure that maximal one message pro second will appear animated, which should be totally fine with 0.3s animation duration. 
It’s actually quite easy, as animation is triggered by the `:new?` property of message, so it will be quite simple to achieve that.

**Comment by flexsurfer** [comment]
@janherich sound reasonable, do you plan to make this changes in this PR?

**Comment by janherich** [comment]
@flexsurfer Yes, I will implement it here, already change status to WIP and moved back to `CONTRIBUTOR` lane.

**Comment by janherich** [comment]
I did easy fix to almost completely solve the animation problem - whenever message is received and the user is having an app open with chat where message belongs, `:appearing?` flag (which controls whether to animate the message appearing) is set to false, so once the chat is later opened (with let's say 50 new messages), it doesn't mean that suddenly 50 messages will be animated at once (which was the case before).
In my testing, it solves 95% of the problems. I was still able to get very occasional super small glitches with chat opened in which lot of messages are being received, but it wasn't really problematic, at least on iPhone SE/iOS simulator when joining the active `#status` channel.
The PR is already growing little bit large, so I would like the QA to test it throughly, share performance findings and hopefully get it merged.
We can then later work on the idea of batched message receives (with only last message being animated) when it proves to be necessary.

File: src/status_im/chat/events.cljs (hunk 2 above)
Comment (yenda) [reply src/status_im/chat/events.cljs:198]: mapv then to keep a vector

**Comment by Serhy** [comment]
@janherich the Android: https://i.diawi.com/uNuWFy and iOS: https://i.diawi.com/qCif4e app versions was tested on Samsung J7 Android 7.0 (real device) and iPhoneX (iOS 11.2.2)

I didn't see any regressions from functional side: messaging between contacts in chats, send/receive Tx works fine. 
Neither I did see regressions on UI performance side. 

The  `783ac71f+` version UI performance similar to the latest develop build. Maybe a bit a bit better.
I tested in following way to compare: 
- Contact A created on Android side.
- Contact B created on iOS side.
- Contact A received 50 messages in GroupChat and 50 messages in 1-1chat with contactB.
- Contact A - re-logins: once on Home screen - navigate from this point to 1-1chat -> Home -> -
- GroupChat ->Home -> Wallet -> Profile -> Home -> 1-1Chat -> ...the lags seen up to 400-500ms. 
- Contact A logs out. While log out - Contact A received 30 messages in Group chat and 30 messages in 1-1chat. 
- Contact A log in: once on Home screen - navigate between different screens. Lags up to 500-600ms. 

What i'm curious is that the same version showed me different results:
So above flow I used twice:
**TF session 1**: https://app.testfairy.com/projects/4803622-status/builds/7599034/sessions/14/?accessToken=8sO3qpf8-AuJAwE6AHXVNeUF8is
**TF session 2**: https://app.testfairy.com/projects/4803622-status/builds/7599034/sessions/15/?accessToken=8-xEDj3pdiVOt1tggFkhy45Grjs

You can check out CPU performance of both sessions: on the TF session 1 see there CPU spikes with the long duration (comparing to TF session 2). This was the after-log-in time and in the first case (TF session 1) - there were really huge lags (up to 5 seconds - could be seen on video).

And latest develop build with the same flow above:
**TF session 3 (develop):** https://app.testfairy.com/projects/4803622-status/builds/7593342/sessions/21/?accessToken=wnv01GfhQwnzQyvEJA7Lx-51STA - pretty similar to TF session 2.

I want to cover some more cases for this branch, however so far from what I covered the UI performance is on the similar stage as we had.

**Comment by janherich** [comment]
Thanks @Serhy , great report !
I can add one case where I found the PR performance to be definitely better then current develop, so maybe you can re-test and confirm that:
Open app and open `#status` public chat while new messages are being received (you can see the unread counter being incremented), wait till you have 50-100 unread messages and open the chat.
On develop, I can see noticeable lag till all the messages are rendered, this is not the case with PR branch.

**Comment by Serhy** [comment]
This has been tested with above approach, (described in https://github.com/status-im/status-react/pull/3160#issuecomment-361714125) and with the case @janherich described. 
Two times on latest develop build and twice with the Android: https://i.diawi.com/PEVV5L and iOS: https://i.diawi.com/YG6Ee1. 
Android 7.0 on Samsung J7 and iPhoneX real devices used. 

When contact received 100 messages while offline and log in - navigating to respective chat (1-1 / Group) on Develop takes ~5-6 seconds for messages to appear on Android 7.0. 
The same case on d479d3c1+ app version on same device showed 1.5-2 seconds delay for received and not read messages to appear on Chat screen. 

TF session for Develop (0.9.10-621-g8f1bc58b+) https://app.testfairy.com/projects/4803622-status/builds/7599983/sessions/11/?accessToken=tW3a630YvpgcRET7-cKeFrYOCLY (from 01:25)
TF session for d479d3c1+ https://app.testfairy.com/projects/4803622-status/builds/7608425/sessions/4/?accessToken=FKAxeFvS60J8WgVMzfMGiqfmYRQ (from 01:50)

We just introduced one regression :( 
The deleted 1-1 chat with messages - re-appear after user re-login to account. Not happening with Group chat. This is not reproduced on develop branch.
TF session where removed 1-1 chat re-appearing with all messages counter incremention: https://app.testfairy.com/projects/4803622-status/builds/7608425/sessions/14/?accessToken=CuBjIrlhFmvxbNPtbZZUMXYjXak

**Comment by janherich** [comment]
Hi @Serhy the introduced regression should be already fixed:
Branch: PR-3160
Android: https://i.diawi.com/6JaZS4
iOS: https://i.diawi.com/wL91Cv

**Comment by Serhy** [comment]
@janherich , thanks a lot!  Above builds: Android https://i.diawi.com/6JaZS4 and
iOS https://i.diawi.com/wL91Cv tested on Android 7.0 and iOS 11.2.2.

The issue with re-appearing 1-1chat still present though, - but behavior a bit different now:
At the moment - if 1-1 chat was deleted before account re-log in, - after contact logs in to account -> 1-1 chat re-appears but no message counter increment happens. (Versus to what we had before: 1-1 chat re-appearing with counter incremention of all the messages contact had in this 1-1 chat)

**To reproduce:**
1) Contact A (Android 7.0) has added Contact B (iOS in my case). 
2) Contact B - tap 'Add in contacts' from 1-1chat with Contact A
3) Contact A and Contact B delete 1-1 chat on both sides
4) Both contacts re-log in and stays on Home screen for 1 minute

**Actual result**: 1-1 chat re-appeared on both sides.

No regressions noticed from performance side.

Additionally I experienced app crash on Android after log in to account with deleted 1-1 chat - but confirmed the same in Develop branch (reported here https://github.com/status-im/status-react/issues/3189)

**Comment by janherich** [comment]
Branch: PR-3160
Android: https://i.diawi.com/tFCTAb
iOS: https://i.diawi.com/e1R7aB

**Comment by janherich** [comment]
Branch: PR-3160
Android: https://i.diawi.com/RbnNGw
iOS: https://i.diawi.com/1yQbTF

**Comment by Serhy** [comment]
Thanks a lot @janherich 
The Android: https://i.diawi.com/RbnNGw and iOS: https://i.diawi.com/1yQbTF app versions were tested on iPhoneX and Android 7.0 (Samsung J7 real device)
 
It looks better now but 1-1 chat still re-appering, with just different steps. Before (in https://github.com/status-im/status-react/pull/3160#issuecomment-362271659)  it was reproducible after first re-login even if no messages were sent in previously deleted 1-1 chat. 

Now, 1-1 chat re-appears when:

1) Open Status on Device 1 and create user (lets call this Contact 1)
2) Open Status on Device 2 and create user (lets call this Contact 2)
3) Contact 1 adds Contact 2 in contacts and sends 10 text messages. 
4) Contact 2 from 1-1 tap 'Add to contacts' and sends 10 text messages
5) Contact 1 and Contact 2 - logs out.
6) Contact 1 and Contact 2 logs in to account and stays for 2-3 minutes on home screen. (**Note: nothing reappeared on this stage**)
7) Contact 1 and Contact 2 - logs out.
8) Contact 1 and Contact 2 logs in to account

**Actual result:** 1-1 chat reappeared on both sides.

TF session representing above steps:
Device 1: https://app.testfairy.com/projects/4803590-status/builds/7618314/sessions/13/?accessToken=A1bYlVo/LW9hldQzc-VGB6azj3w (the last login with 1-1 chat reappering from 06:57)
Device 2: https://app.testfairy.com/projects/4803622-status/builds/7624261/sessions/19/?accessToken=hBD0mDi6pM84p/2AXAup3rscq44 (the last login with 1-1 chat reappering from 06:52)

This I could not reproduce on latest Develop.

**Comment by janherich** [comment]
@Serhy I did my utmost to fix the bug and tested extensively, fingers crossed:
Branch: PR-3160
Android: https://i.diawi.com/Cn6VM3
iOS: https://i.diawi.com/nQ4oqB

**Comment by oskarth** [comment]
Great stuff @janherich!

**Comment by Serhy** [comment]
Latest builds Android: https://i.diawi.com/WqhQdJ and iOS: https://i.diawi.com/7dPch4 tested on Samsung J7 - Android 7.0) and iPhoneX (iOS 11.2.2). 
All regressions above are fixed! As well as the performance results improvement kept: for 100 received messages it takes ~1.5-2 seconds to load for above builds vs 5-5.5 seconds I see on build from develop branch for same device (Android 7.0 in my case). 
Awesome job, @janherich! Lets merge!
